File: week4_paths2/2_negative_cycle/negative_cycle.py
#Uses python3

import sys


def negative_cycle(adj, cost):
    #write your code here
    n = len(adj)
    cycle = 0
    dists = [10**7]*n
    dists[0] = 0
    for i in range(n+1):
        cycle = 0
        for i in range(n):
            for j,w in enumerate(adj[i]):
                if dists[w]> dists[i]+cost[i][j]:
                    dists[w] = dists[i] + cost[i][j]
                    cycle = 1
        if cycle==0:
            return 0
    return cycle


if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(zip(data[0:(3 * m):3], data[1:(3 * m):3]), data[2:(3 * m):3]))
    data = data[3 * m:]
    adj = [[] for _ in range(n)]
    cost = [[] for _ in range(n)]
    for ((a, b), w) in edges:
        adj[a - 1].append(b - 1)
        cost[a - 1].append(w)
    print(negative_cycle(adj, cost))

File: week4_paths2/2_negative_cycle/negative_cycle.cpp
#include <iostream>
#include <vector>

using std::vector;

int negative_cycle(vector<vector<int> > &adj, vector<vector<int> > &cost) {
  //write your code here
  int n=adj.size();
  vector<int> dists(n,10000000);
  dists[0] = 0;
  int cycle=0;
  for(int k=0;k<n+1;++k){
    cycle = 0;
    for(int i=0;i<n;++i)
      for(int j=0;j<adj[i].size();++j)
        if(dists[adj[i][j]] > dists[i]+cost[i][j]) { dists[adj[i][j]] = dists[i]+cost[i][j]; cycle=1;}
    if(cycle==0) return 0;
  }
  return cycle;
}

int main() {
  int n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  vector<vector<int> > cost(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y, w;
    std::cin >> x >> y >> w;
    adj[x - 1].push_back(y - 1);
    cost[x - 1].push_back(w);
  }
  std::cout << negative_cycle(adj, cost);
}
